controllers/machine_controller.go

```diff
@@ -534,6 +541,41 @@ func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cl
 	return ctrl.Result{}, nil
 }
 
+// pod deletion and volume detach happen asynchronously, so pod could be deleted before volume detached from the node
+// this could cause issue for some storage provisioner, for example, vsphere-volume this is problematic
+// because if the node deleted before detach success, then the underline VMDK will be deleted together with the Machine
+// so wait for volume detach from the node after node draining.
+func (r *MachineReconciler) waitForVolumeDetach(ctx context.Context, cluster *clusterv1.Cluster, nodeName string, machineName string) error {
+	log := ctrl.LoggerFrom(ctx, "cluster", cluster.Name, "node", nodeName, "machine", machineName)
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		log.Error(err, "Error creating a remote client for cluster while waiting for volume detach")
+		return err
+	}
+
+	key := types.NamespacedName{Name: nodeName}
+	node := &corev1.Node{}
+
+	waitErr := wait.PollImmediate(2*time.Second, 20*time.Second, func() (bool, error) {
```

Comment:
I'm not sure if we should block exection here for 20 seconds, instead of just returning and waiting a notification from the Node watch we already have in this controller?